src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
-    if (ret != 0) {
-        if (x != NULL && localDh != *x)
-            XFREE(localDh, NULL, DYNAMIC_TYPE_OPENSSL);
-        localDh = NULL;
-        goto end;
+    if (ret == 1) {
+        /* On success record that the internal values have been set. */
+        dh->inSet = 1;
     }
 
-    if (x != NULL)
-        *x = localDh;
+    return ret;
+}
+
+/* Get the size, in bytes, of the DH key.
+ *
+ * Return code compliant with OpenSSL.
+ *
+ * @param [in] dh  DH key.
+ * @return  -1 on error.
+ * @return  Size of DH key in bytes on success.
+ */
+int wolfSSL_DH_size(WOLFSSL_DH* dh)
+{
+    int ret = -1;
+
+    WOLFSSL_ENTER("wolfSSL_DH_size");
 
-    /* Put p and g in as big numbers. */
-    if (localDh->p != NULL) {
-        wolfSSL_BN_free(localDh->p);
-        localDh->p = NULL;
+    /* Validate paramater. */
+    if (dh != NULL) {
+        /* Size of key is size of prime in bytes. */
+        ret = wolfSSL_BN_num_bytes(dh->p);
     }
-    if (localDh->g != NULL) {
-        wolfSSL_BN_free(localDh->g);
-        localDh->g = NULL;
+
+    return ret;
+}
+
+/**
+ * Return parameters p, q and/or g of the DH key.
+ *
+ * @param [in]  dh  DH key to retrieve parameters from.
+ * @param [out] p   Pointer to return prime in. May be NULL.
+ * @param [out] q   Pointer to return order in. May be NULL.
+ * @param [out] g   Pointer to return generator in. May be NULL.
+ */
+void wolfSSL_DH_get0_pqg(const WOLFSSL_DH *dh, const WOLFSSL_BIGNUM **p,
+    const WOLFSSL_BIGNUM **q, const WOLFSSL_BIGNUM **g)
+{
+    WOLFSSL_ENTER("wolfSSL_DH_get0_pqg");
+
+    if (dh != NULL) {
+        /* Return prime if required. */
+        if (p != NULL) {
+            *p = dh->p;
+        }
+        /* Return order if required. */
+        if (q != NULL) {
+            *q = dh->q;
+        }
+        /* Return generator if required. */
+        if (g != NULL) {
+            *g = dh->g;
+        }
     }
-    localDh->p = wolfSSL_BN_bin2bn(p, pSz, NULL);
-    localDh->g = wolfSSL_BN_bin2bn(g, gSz, NULL);
-    if (localDh->p == NULL || localDh->g == NULL) {
-        if (x != NULL && localDh != *x)
-            wolfSSL_DH_free(localDh);
-        localDh = NULL;
+}
+
+#if !defined(HAVE_FIPS) || (defined(HAVE_FIPS) && !defined(WOLFSSL_DH_EXTRA)) \
+ || (defined(HAVE_FIPS_VERSION) && FIPS_VERSION_GT(2,0))
+#if defined(OPENSSL_ALL) || \
+    defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x10100000L
+/* Sets the parameters p, g and optionally q into the DH key.
+ *
+ * Ownership of p, q and g get taken over by "dh" on success and should be
+ * free'd with a call to wolfSSL_DH_free -- not individually.
+ *
+ * @return  1 on success.
+ * @return  0 on failure.
+ */
+int wolfSSL_DH_set0_pqg(WOLFSSL_DH *dh, WOLFSSL_BIGNUM *p,
+    WOLFSSL_BIGNUM *q, WOLFSSL_BIGNUM *g)
+{
+    int ret = 1;
+
+    WOLFSSL_ENTER("wolfSSL_DH_set0_pqg");
+
+    /* Validata parameters - q is optional. */
+    if ((dh == NULL) || (p == NULL) || (g == NULL)) {
```

Comment:
This isn't quite how OpenSSL handles the parameters. From their code:

>    /* If the fields p and g in d are NULL, the corresponding input
>     * parameters MUST be non-NULL.  q may remain NULL.
>     */

Reply:
Changed.
Added tests.